### ISSUE TITLE
Multiple StorageClasses

### ISSUE THREAD

**Issue by temujin9** [cybozu-go/topolvm#125]
## What

Some use cases of TopoLVM are on big data systems with SSDs. It would be nice to support them as a separate StorageClass, so that disk performance can be selected for.

## How

Add a parameter to StorageClass that sets the target volume group. That might also free from the need to specify it via the systemd startup line, although it might be prudent to have an allow-list / block-list to prevent messing with system volume groups.

## Checklist

- [x] Finish implementation of the issue
- [x] Test all functions
- [x] Have enough logs to trace activities
- [x] Notify developers of necessary actions

**Comment by deveshk0** [comment]
Much needed enhancement.

**Comment by deveshk0** [comment]
@ymmt2005 Thanks for taking this up

**Comment by ymmt2005** [comment]
@deveshk0 You are welcome!
We are considering this seriously now as there seems a handful of potential users.

**Comment by leonenkoa** [comment]
I think if we consider to support multiple volume groups with separate StorageClass, makes sense to create volume group for specific StorageClass, what do you think ?

**Comment by ymmt2005** [comment]
Do you mean one volume group for one StorageClass?
I don't think so because StorageClass has other options such as filesystem type.

**Comment by leonenkoa** [comment]
So we should bind somehow volume groups,filesystems and StorageClasses and I don't have any other idea except to provision all this parts with TopoLVM i.e. with some other service.

**Comment by deveshk0** [comment]
We have come up with something like this - 

```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: topolvm-sdd
provisioner: topolvm.cybozu.com
parameters:
  "csi.storage.k8s.io/fstype": "xfs"
  "topolvm.cybozu.com/vol-group": "ssd-group"
volumeBindingMode: WaitForFirstConsumer
allowVolumeExpansion: true
```

```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: topolvm-hdd
provisioner: topolvm.cybozu.com
parameters:
  "csi.storage.k8s.io/fstype": "xfs"
  "topolvm.cybozu.com/vol-group": "hdd-group"
volumeBindingMode: WaitForFirstConsumer
allowVolumeExpansion: true
```

This parameters are available  while creating the volume.

**Comment by deveshk0** [comment]
> So we should bind somehow volume groups,filesystems and StorageClasses and I don't have any other idea except to provision all this parts with TopoLVM i.e. with some other service.

Agreed. There should be option to add vol-group, because user can create multiple storage classes  with same vol-group but diff `fstype`

**Comment by zoetrope** [comment]
Thank you for your suggestions.
I think there are many considerations to implement this feature.
So I'll start writing a design document for discussion.

**Comment by ymmt2005** [comment]
targetted for 0.5

**Comment by deveshk0** [comment]
@ymmt2005 Any planned date for the 0.5 Milestone?

**Comment by ymmt2005** [comment]
Not particularly.

**Comment by deveshk0** [comment]
@zoetrope Can you create a beta build for this since all test are passing now.

**Comment by zoetrope** [comment]
@deveshk0 
I uploaded container image: quay.io/cybozu/topolvm:0.5.0-beta1
This is still in beta so the specs may change.

**Comment by zoetrope** [comment]
@temujin9 @deveshk0 @leonenkoa 
I wrote a design document about multiple volume groups.
https://github.com/cybozu-go/topolvm/pull/131

I have two proposals on `How to annotate resources`.
There are pros and cons in each proposal, I would like to adopt `1) insert multiple resources`.
How about you?

**Comment by leonenkoa** [comment]
@zoetrope I'm also adopt first approach.I think the cons of the second approach is quite important to ignore them.

**Comment by deveshk0** [comment]
@zoetrope I think first approach makes more sense.

**Comment by zoetrope** [comment]
@leonenkoa @deveshk0 
Thank you for your opinion.
We decided to adopt option 1 and I started implementing.

**Comment by zoetrope** [comment]
I said I decided to adopt option 1.
However, it turned out that option 1 causes harder settings than I thought.

I show pros/cons again.

Option 1:
Pros:
- The weight of extender can be adjusted for each volume group.
Cons:
- The settings of scheduler policy are complicated and must be rewritten according to your environment.

Option 2:
Pros:
- The settings of scheduler policy is simple.
- For pods that use only one volume group, scheduling can be done as usual.
- For pods that use two or more volume group, scheduling to a node with insufficient capacity can be avoided.
Cons:
- The weight of extender cannot be adjusted individually when using two or more volume group.

In most cases, option 2 works without problems, so I'll implement it with option2 first.
If you want to set individual weights, you can create a new issue.

**Comment by ymmt2005** [comment]
@zoetrope Please update the design proposal document.

**Comment by ymmt2005** [comment]
I talked directly with @zoetrope that we should not expose volume group names
in user-visible resources.  This is because LVM volume group names are node-local,
and the current TopoLVM implementation does not expose VG names.

Instead, introducing a new concept to represent the class of storage devices
and mapping the classes to VG names in `lvmd` seems a better idea.

@zoetrope is going to revise the proposed design.

**Comment by ymmt2005** [comment]
We have finally merged the multi-VG support.

We went through a number of design revisions to ensure backward-compatibility,
but in the end, I think we got it right.

The next step is to release a beta and test upgrade path.

**Comment by ymmt2005** [comment]
We verified the migration path from previous installations, and
released v0.5 with an upgrade guide.
https://github.com/cybozu-go/topolvm/releases/tag/v0.5.0

Enjoy!
